src/ripple/nodestore/impl/DeterministicShard.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2020 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/app/main/Application.h>
+#include <ripple/core/ConfigSections.h>
+#include <ripple/nodestore/Manager.h>
+#include <ripple/nodestore/impl/DeterministicShard.h>
+#include <ripple/nodestore/impl/Shard.h>
+#include <ripple/protocol/digest.h>
+#include <fstream>
+#include <nudb/detail/format.hpp>
+#include <nudb/nudb.hpp>
+#include <openssl/ripemd.h>
+
+namespace ripple {
+namespace NodeStore {
+
+DeterministicShard::DeterministicShard(
+    Application& app,
+    DatabaseShard const& db,
+    std::uint32_t index,
+    uint256 const& lastHash,
+    beast::Journal j)
+    : inited_(false)
+    , nodeset_([](nodeptr l, nodeptr r) { return l->getHash() < r->getHash(); })
+    , app_(app)
+    , db_(db)
+    , index_(index)
+    , hash_(hash(lastHash))
+    , tempdir_(db.getRootDir() / (std::to_string(index_) + ".tmp"))
+    , finaldir_(db.getRootDir() / std::to_string(index_))
+    , ctx_(std::make_unique<nudb::context>())
+    , j_(j)
+{
+}
+
+DeterministicShard::~DeterministicShard()
+{
+    close(true);
+}
+
+uint160
+DeterministicShard::hash(uint256 const& lastHash) const
+{
+    RIPEMD160_CTX c;
+    uint8_t result[20];
+
+    RIPEMD160_Init(&c);
+    RIPEMD160_Update(&c, &lastHash, sizeof(lastHash));
+    RIPEMD160_Update(&c, &index_, sizeof(index_));
+    auto firstSeq = db_.firstLedgerSeq(index_);
+    RIPEMD160_Update(&c, &firstSeq, sizeof(firstSeq));
+    auto lastSeq = db_.lastLedgerSeq(index_);
+    RIPEMD160_Update(&c, &lastSeq, sizeof(lastSeq));
+    RIPEMD160_Update(&c, &Shard::version, sizeof(Shard::version));
+    RIPEMD160_Final(result, &c);
+    return uint160::fromVoid(result);
+}
+
+std::uint64_t
+DeterministicShard::digest(int n) const
+{
+    int ind;
+    std::uint64_t res = 0;
+    for (int i = sizeof(std::uint64_t) - 1; i >= 0; --i)
+    {
+        res <<= 8;
+        ind = i + n * sizeof(std::uint64_t);
+        if (ind < 160 / 8)
+            res += hash_.data()[ind] & 0xff;
+    }
+    return res;
+}
```

Comment:
This function works but seems a little overly complicated for what it's trying to do. You don't need to change it, but I'm wondering how you feel about this:

```
std::uint64_t
DeterministicShard::digest(int n) const
{
    auto const data = hash_.data();

    if (n == 2)
    { // Extract 32 bits:
        return (static_cast<std::uint64_t>(data[19]) << 24) +
            (static_cast<std::uint64_t>(data[18]) << 16) +
            (static_cast<std::uint64_t>(data[17]) << 8) +
            (static_cast<std::uint64_t>(data[16]));
    }

    std::uint64_t ret = 0;

    if (n == 0 || n == 1)
    { // Extract 64 bits 
        for(int i = n; i < 16; i += 2)
            ret = (ret << 8) + data[i];
    }

    return ret;
}
```

Note that `digest(2)` now returns 32 bits (as a 64-bit value), while `digest(0)` and `digest(1)` return 64-bit values.

Reply:
Done